torch.sort: Optimize memory usage with downcasting and upcasting indices for functional variant and direct type for out-variant (Reopen)

voidbag opened the pull request:
Fixes [#147628](https://github.com/pytorch/pytorch/issues/147628)
Reopen [#147629](https://github.com/pytorch/pytorch/pull/147629)

The static dtype of indices was kLong(64bit), consuming exccessive memory.
It reduces memory usage, determining the dtype of indices dynamically.
The dtype is one of Byte, UInt16, UInt32 and UInt64.

- This PR makes at::linspace support uint16, uint32 and uint64



- 1, 2, 4 value types are added in cub-RadixSortPairs-*.cu files for parallel compile suggested in [148936](https://github.com/pytorch/pytorch/pull/148936)
- @ngimel If the parallel compile takes too long to compile the source code in win build, I'll remove one-by-one from 8 bits to 16 bits.


Recap: Performance Report ([#147629](https://github.com/pytorch/pytorch/pull/147629))
```
shape = (4019, 87585) or (69878, 10677)
# indices_dtype = (None|torch.int32|torch.int16) # only in wip-dynamic-dtype-sort-voidbag-future, not current branch 
# ret = torch.sort(torch.zeros(shape, dtype=torch.bool, device="cuda:0"), indices_dtype=indices_dtype)
ret = torch.sort(torch.zeros(shape, dtype=torch.bool, device="cuda:0"))

# (Peak Mem Usage(MiB))
# !nvidia-smi

torch.cuda.empty_cache()
# (Mem Usage W/O cache)
# !nvidia-smi

```
|    | shape          | method                    |   Peak Mem Usage(MiB) |   Mem Usage W/O cache(MiB) |
|---:|:---------------|:--------------------------|----------------------:|---------------------------:|
|  0 | (4019, 87585)  | default                   |                  9372 |                       3328 |
|  1 | (4019, 87585)  | downcast->upcast(this PR)          |                  8032 |                       3328 |
|  2 | (4019, 87585)  | indices_dtype=torch.int32 |                  5346 |                       1986 |
|  3 | (69878, 10677) | default                   |                 19522 |                       6712 |
|  4 | (69878, 10677) | downcast->upcast(this PR)          |                 12408 |                       6712 |
|  5 | (69878, 10677) | indices_dtype=torch.int16 |                  6714 |                       2442 |



cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @aditew01

voidbag committed:
Sort: optimize CPU/GPU memory usage with dynamic dtype of indices

The static dtype of indices was kLong(64bit), consuming exccessive memory.

This feature reduces memory usage, determining the dtype of indices dynamically.

The dtype is one of at::ScalarType(s) Byte, UInt16, UInt32 and UInt64.

Signed-off-by: Taeksang Kim <voidbag@gmail.com>

voidbag committed:
linspace: support UInt16, UInt32, UInt64

It's for dynamic dtype of at::sort.

Signed-off-by: Taeksang Kim <voidbag@gmail.com>

voidbag committed:
sort: add 'dynamic_indices_dtype'

The dtype of indices is selected among uint8, 16, 32 and 64 automatically, if the option is set.

Signed-off-by: Taeksang Kim <voidbag@gmail.com>

voidbag committed:
sort: handle onnx test failures

Handle Onnx Test Failure
- Handles onnx test failure by adding parameters in
  torch/onnx/symbolic_opset(9|10|11).py

Handle Backwards Compatibility Issue
- Tensor.sort(true, 0, true) in torchvision calls
  Tensor.sort(int dim, bool descending, bool dynamic_indices_type)
  rather than Tensor.sort(optional<bool> stable, int dim, bool descending)
- It's resolved by changing the function's prototype
  as sort(int dim, bool descending, ScalarType indices_dtype,
          bool dynamic_indices_type)

Add `str == "at::kLong"` in python_arg_parser.cpp.
- `at::kLong` cannot be parsed in
  `python_(variable_methods|torch_functionsEverything|torch_functions_2)`.cpp
   in `autograd/generated` directory
- It's resolved by adding "at::kLong" or-condition to "torch.int64"

Signed-off-by: Taeksang Kim <voidbag@gmail.com>

voidbag committed:
torch.sort: indices supports all integral dtypes

change dispatch macro to support all integral types of indices

Signed-off-by: Taeksang Kim <voidbag@gmail.com>

voidbag committed:
Revert to origin signature of torch.sort

Revert to origin signature of torch.sort...

Signed-off-by: Taeksang Kim <voidbag@gmail.com>

voidbag committed:
torch.sort: downcast indices and upcast to kLong

The process of downcasting and upcasting indices saves memory usage.
It doesn't require the signature of torch.sort.

Signed-off-by: Taeksang Kim <voidbag@gmail.com>

voidbag committed:
SortingKernel.cpp: fix win build failure

remove `#ifdef in AT_DISPATCH_V2` for win build

Signed-off-by: Taeksang Kim <voidbag@gmail.com>

voidbag committed:
temporary commit to prevent bot from closing it

Signed-off-by: Taeksang Kim <voidbag@gmail.com>

voidbag committed:
Add (1|2|4) bytes values to cub-RadixSortPairs

- remove and merge cub-RadixSortPairs-*.cu

Signed-off-by: Taeksang Kim <voidbag@gmail.com>
Signed-off-by: Taeksang Kim <ts.kim@hyperaccel.ai>

voidbag commented:
Compilation Time Report
@ngimel 

Hardware Setting
- CPU: 9950x (16 cores, 32 threads)
- memory: 92GB
- GPU: 5070ti
Software Setting
- Host OS: Ubuntu 24.04.2 LTS
- docker image: pytorch/pytorch:2.8.0-cuda12.8-cudnn9-devel
- main hash: 6493f23b9931ff9661a36cfb3752d1b528467805

Compilation Time in main branch
```
time MAX_JOBS=16 python -m pip install --no-build-isolation -v -e .
real    24m54.761s
user    379m34.279s
sys     20m32.588s
```
Compilation Time in this PR branch
```
sudo bash -c "echo 3 > /proc/sys/vm/drop_caches" #In host OS

time MAX_JOBS=16 python -m pip install --no-build-isolation -v -e .
real    28m9.156s
user    401m36.331s
sys     21m1.119s
```


I'll try first to remove unnessary dtypes in `AT_DISPATCH_V2` in this patch, to include only uint8_t, uint16_t, uint32_t, int64_t.

For example, SortStable.cu in this patch contains too many dtypes using **AT_INTEGRAL_TYPES_V2**.
https://github.com/pytorch/pytorch/blob/e0b9368bdea8e91e5f7d8d67c7fffff910dabf90/aten/src/ATen/native/cuda/SortStable.cu#L293-L303

Skylion007 commented:
Does this improve benchmark performance too btw? Sort is pretty important perf bottleneck for MoE router, so I would advocate for merging it despite the compile time increase if it made things significantly faster.

I'd also like to point out that just adding uint32 support here would greatly reduce memory and may accelerate things in the common case. It's very likely to have less than 4.2B elements that you are sorting. The cases of 65k and 256 elements are way rarer and likely not huge memory and perf bottlenecks. You can unlock most of the perf improvements likely by using the canUse32bit like dispatch on just the dimension and it would vastly simplify this PR.

Skylion007 commented on aten/src/ATen/cuda/cub-RadixSortPairs-f16s.cu:
```diff
@@ -0,0 +1,11 @@
+
+namespace at::cuda::cub::detail {
+
+AT_INSTANTIATE_SORT_PAIRS(c10::BFloat16, 8)
```
So these are we all separated out into one instantiation per CUDA file in a previous PR to improve compile time @voidbag . Re-aggregating them like this might be partially responsible for the compile time regression since it makes the instatiation less parallel.

Since we have a lot of inationation of similar bit width now, I am not sure if will increase compile size if we do so btw.

Skylion007 commented:
Due to a bug, we can't support byte indexing on most operators. https://github.com/pytorch/pytorch/issues/61819 not sure why this didn't fail here, but it makes me suspicious of the tests.

voidbag committed:
Remove unused indices_dtypes

- The indices_dtypes are kByte, kUInt16, kUInt32, kLong

Signed-off-by: Taeksang Kim <voidbag@gmail.com>
Signed-off-by: Taeksang Kim <ts.kim@hyperaccel.ai>

voidbag committed:
Replace linspace with arange

- index generation is done through arange, instead of linspace.

Signed-off-by: Taeksang Kim <voidbag@gmail.com>
Signed-off-by: Taeksang Kim <ts.kim@hyperaccel.ai>

voidbag commented:
[Almost Final Compile Time Report]
The previous report doesn't catch **other misuses `AT_EXPAND(AT_INTEGRAL_TYPES_V2) in Sort.cu and SortingKernel.cu`** of AT_DISPATCH_V2. The previous report just **cached one misuse in `SortStable.cu`**. It was my mistake.

4-types(kByte, kUInt16, kUInt32, kLong) support looks feasible for me, after I have modified all `AT_EXPAND(AT_INTEGRAL_TYPES_V2)` misuses.

Experiment environment of this report is same with that of the previous report (**16 cores 32 threads, MAX_JOBS=16**)
1. Summary
    1. Compile Time (Intact Main branch vs 4-types Support)
            - **real:   24m57.773s -> 25m34.410s (looks feasible)**
            - user:  379m2.437s -> 388m53.027s
            - sys:      20m28.529s -> 20m53.763s
    2. Compile Time (Aggregate cub-RadixSortPairs-*-*.cu vs 4-types Support)
            - **real:   24m56.632s -> 25m34.410s (looks feasible)**
            - user:  381m42.330s-> 388m53.027s
            - sys:      20m49.952s -> 20m53.763s
    3. Compile Time (2-types Support vs 4-types Support)
            - **real:   25m10.666s -> 25m34.410s (looks feasible)**
            - user:  382m26.211s-> 388m53.027s
            - sys:      20m36.908s -> 20m53.763s


2. Detailed Report
    1. main branch(intact)
        ```
        real    24m57.773s
        user    379m2.437s
        sys     20m28.529s
        ```
    2. Aggregate cub-RadixSortPairs-*.cu (Acceptable degradation)
        I think the main bottleneck can be `cub-RadixSortPairs-scalars.cu`, which contains `AT_FORALL_SCALAR_TYPES_AND2(Bool, Half, AT_INSTANTIATE_SORT_PAIRS_8)`. The other files just contain single (type, size) pair. So I decided to merge them for readability. I think it's better to parallelize `cub-RadixSortPairs-scalars.cu`.
        - cub-RadixSortPairs-{f16,uint16, uint32, uint64}-8 to single file
        - cub-RadixSortPairs-int32-{1,2,4} to single file
        - cub-RadixSortPairs-int64-{1,2,4} to single file
        ```
        real    24m56.632s
        user    381m42.330s
        sys     20m49.952s
        ```
   3. single-type(kLong) support(AT_DISPATCH_V2 includes only kLong)
       ```
       real    25m0.443s
       user    379m20.002s
       sys     20m53.763s
       ```
    4. two-types(kLong, kUInt32) support
        ```
        real    25m10.666s
        user    382m26.211s
        sys     20m36.908s
        ```
    5. 4-types(kLong, kUInt32, kUInt16, kByte) support
        ````
        real    25m0.443s
        user    379m20.002s
        sys     20m53.763s
        ````

voidbag commented:
@Skylion007 
I make it use arange instead of linspace.

voidbag commented:
> Due to a bug, we can't support byte indexing on most operators. #61819 not sure why this didn't fail here, but it makes me suspicious of the tests.

I think the reason is that the indices, which is converted into byte, uint16, etc dtype, is just used as value of sort. Even the converted indices value is used temporarily, not to be used for actual indexing



---
Eventually, I want to add optional indices_dtype parameter into sort op signature, which doesn't require upcast to kLong.

> |    | shape          | method                    |   Peak Mem Usage(MiB) |   Mem Usage W/O cache(MiB) |
> |---:|:---------------|:--------------------------|----------------------:|---------------------------:|
> |  4 | (69878, 10677) | downcast->upcast(this PR)          |                 12408 |                       6712 |
> |  5 | (69878, 10677) | indices_dtype=torch.int16 |                  6714 |                       2442 |

In device with small memory, large (kLong) indices can be a burden for small device...

voidbag commented on aten/src/ATen/native/cuda/SortStable.cu:
```diff
@@ -288,4 +290,17 @@ void launch_stable_sort_kernel(
+  AT_DISPATCH_V2(indices.scalar_type(), "launch_stable_sort_kernel", AT_WRAP([&] {
+    scalar_t* indices_ptr = indices.mutable_data_ptr<scalar_t>();
+    do_launch_stable_sort_kernel(self, dim, descending, values, indices_ptr);
+  }), kByte, kUInt16, kUInt32, kLong);
```
You can change the `kByte, kUInt16, kUInt32, kLong` into `kUInt16, kLong` or `kLong` for testing

voidbag replied:
Plus, you should update `cub-RadixSortPairs-*.cu` for exact testing, removing unnecessary type.

voidbag reviewed (commented):
Additional Information for Testing

Skylion007 commented on aten/src/ATen/native/Sorting.cpp:
```diff
@@ -100,7 +100,7 @@ void _fill_indices(const TensorBase &indices, int64_t dim) {
   assert(0 <= dim && dim < ndim);
   auto dim_size = indices.size(dim);
-  auto idx_dim = at::arange(0, dim_size, indices.options().dtype(at::kLong));
+  auto idx_dim = at::arange(0, dim_size, indices.options());
```
This still needs a dtype long if the dimsize is large enough, doesn't it?

Skylion007 replied:
Otherwise, it will not float64 for arange incremental kernel and may cause inaccuracy issues above 14 million elements

Skylion007 commented on prevent-closed-by-bot.txt:
delete this obv

Skylion007 reviewed (commented):
Looks like the arange overload is an issue. On GPU, I think they might be implemented as roughly the same though.

Skylion007 commented on aten/src/ATen/native/Sorting.cpp:
```diff
@@ -951,7 +951,29 @@ TORCH_IMPL_FUNC(sort_stable_out)
+    sort_stub(self.device().type(), self, values, *indices_tmp, dim, descending, stable.value_or(false));
+
+    if (!indices_tmp->is_same(indices)) {
+      indices.copy_(*indices_tmp);
```
Any benchmarks to show this copy doesn't massively slow things down? I'm guessing not, but would be good to have some speed improvements in addition to this. Also, how the hell is copy being used a const ref? Doesn't it mutate state?

voidbag replied:
1. Let me have some time to think about performance improvement..

2. const ref tensor.copy_
https://github.com/pytorch/pytorch/blob/2055b66006040424bf6ffeeec8710f1a42244d08/aten/src/ATen/native/Sorting.cpp#L941-L948

    Const ref tensor.copy_ in 948 line is already called in the existing code. I'll think about it too.

voidbag committed:
Add uInt16 and uint32 into arange_cpu

- add uint16, uint32 into arange_cpu
- fix code alignment

Signed-off-by: Taeksang Kim <voidbag@gmail.com>
Signed-off-by: Taeksang Kim <ts.kim@hyperaccel.ai>

voidbag committed:
Use std::signed_t instead of custom template

This patch has been reviewed by @ngimel

Signed-off-by: Taeksang Kim <voidbag@gmail.com>
Signed-off-by: Taeksang Kim <ts.kim@hyperaccel.ai>
Co-authored-by: Natalia Gimelshein <ngimel@meta.com>

voidbag committed:
Add output indices dtype in Meta Function

- with indices_dtype option, you can have workaround way
    ```
    data : Tensor
    result = torch.empty((0,), device=data.device, dtype=data.dtype)
    indices = torch.empty((0,), device=data.device, dtype=torch.uint16)
    torch.sort(data, out=(result, indices))
    ```

Signed-off-by: Taeksang Kim <ts.kim@hyperaccel.ai>

voidbag committed:
torch.sort: use out's dtype as indices_dtype

1. Use out's dtype as indices_dtype.
2. acceptable indices_dtype: (kChar, kByte, kShort, kUInt16, kInt,
   kLong).
3. If kLong is set as out's dtype, then downcast and upcast process is
   executed.

Signed-off-by: Taeksang Kim <voidbag@hyperaccel.ai>

voidbag committed:
Split files to reduce kernel bin for speedup

Signed-off-by: Taeksang Kim <ts.kim@hyperaccel.ai>

voidbag committed:
Optimize SortInplace

In benchmark, down-upcast has no benefits for sortInplace of cuda.
The criterion is 4096

Signed-off-by: Taeksang Kim <ts.kim@hyperaccel.ai>

voidbag committed:
Revert topk modification

Signed-off-by: Taeksang Kim <ts.kim@hyperaccel.ai>

voidbag commented:
@skylion007 and @ngimel:

I’ve performed local build benchmarks (MAX_JOBS=16) to measure the impact. To minimize the 'compile-time that @ngimel mentioned, I implemented kernel sharing for dtypes of the same size (e.g., uint16 and int16 use the same kernel instantiation).

Build Time Comparison:
| Version | User Time | Real Time | Delta |
| :--- | :--- | :--- | :--- |
| **Origin** | 379m 20.1s | **24m 50.3s** | - |
| **This Patch** | 392m 15.8s | **25m 44.0s** | **+53.7s (~3.6%)** |

The increase is roughly 54 seconds. Given the memory efficiency gains for projects like Torchtitan, I believe this is a worthwhile trade-off. Could you please trigger the CI to verify these numbers in the official environment?

voidbag commented:
I need to check the result of ci triggered before this latest CI. I remember that the previous ci triggered succeeded (4 dtypes down up cast only). But I just saw ci success and didn't check the build time...

Except splitting kernel files, i think everything about compile time should be same with the previous one which only does down up cast about 4 dtypes.... If the previous succeeded CI has less compile time, there might be good signal.

ngimel commented:
TopK on cuda is never dispatched to sort, especially with small K

voidbag committed:
Merge Files for each dtypes

Signed-off-by: Taeksang Kim <ts.kim@hyperaccel.ai>

voidbag committed:
Remove prevent-closed-by-bot.txt

Signed-off-by: Taeksang Kim <ts.kim@hyperaccel.ai>

voidbag committed:
Fix HAS_WARP_MERGE_SORT always-false (small stable sort crash on CUDA)

SortDecl.cuh defined HAS_WARP_MERGE_SORT() as (CUDA_VERSION >= 110600), which is
always false for two reasons:
  1. 110600 is a typo for 11060 (CUDA 11.6), the value already used in SortUtils.cuh.
     CUDA 13.0 has CUDA_VERSION 13000, and 13000 < 110600.
  2. SortDecl.cuh included cuda_runtime_api.h, which defines CUDART_VERSION but not
     CUDA_VERSION, so CUDA_VERSION was undefined in Sort.cu's translation unit.

With HAS_WARP_MERGE_SORT() false in Sort.cu but true in the kernel translation units
(SortInplace-*.cu via SortImpl.cuh -> cub.cuh -> cuda.h, plus the correct 11060 in
SortUtils.cuh), sortKeyValueInplace dispatched sort_size <= 128 to MediumRadixSort,
which itself asserts that size <= 128 is handled by a different algorithm:
  INTERNAL ASSERT FAILED at SortImpl.cuh
  "Expected size <= 128 to be handled by a different algorithm"
Every stable sort with sort-dim <= 128 crashed (e.g. torch.sort(x, stable=True) on a
[B, n<=128] tensor), which is exercised by torchtitan MoE training.

Fix: include cuda.h (defines CUDA_VERSION) and correct the threshold to 11060, matching
SortUtils.cuh. This is a header-only change, so the file-splitting binary-size
optimization is preserved.

Test Plan:
  Built torch with this fix (CUDA 13.0, RTX 5070 Ti, driver 580.159.04).

  python3 -c "import torch
  for n in [2, 8, 16, 32, 64, 128, 129, 256, 2048, 4097]:
      for d in (True, False):
          torch.sort(torch.randn(8, n, device='cuda'), dim=1, stable=True, descending=d)
          torch.cuda.synchronize()
  print('all stable sorts ok')"

  Before the fix: RuntimeError INTERNAL ASSERT at SortImpl.cuh for all n <= 128.
  After the fix: all ok, and torchtitan llama4 MoE single-GPU trains to completion.

This change was authored with the assistance of an AI coding assistant.

Signed-off-by: Taeksang Kim <voidbag@gmail.com>
